File: stages/2_download_recordings/download_recordings.js
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const xlsx = require('xlsx');
require('dotenv').config({ path: process.env.CONFIG_PATH ? path.dirname(process.env.CONFIG_PATH) + '/.env' : '../../.env' });

// Load configuration
const configPath = process.env.CONFIG_PATH || path.join(__dirname, '../../config/pipeline_config.json');
const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));

// Setup directories
const baseDir = path.join(__dirname, '../..');
const callIdsDir = path.join(baseDir, 'output/call_ids');
const audioDir = path.join(baseDir, 'output/audio');

// Create audio directory if it doesn't exist
if (!fs.existsSync(audioDir)) {
  fs.mkdirSync(audioDir, { recursive: true });
}

async function getRecordingInfo(callId, accessToken) {
  try {
    const url = `https://integrate.versature.com/api/recordings/call_ids/${callId}/`;
    const response = await axios.get(url, {
      headers: {
        'Accept': 'application/vnd.integrate.v1.4.0+json',
        'Authorization': `Bearer ${accessToken}`
      }
    });
    return response.data;
  } catch (error) {
    if (error.response?.status === 401) {
      throw new Error('Authentication failed - token may have expired');
    }
    throw error;
  }
}

async function downloadAudio(recordingUrl, callId, brokerId, accessToken) {
  try {
    console.log(`🔽 Getting recording info for call ${callId}...`);
    
    // First get the recording info to get the actual download URL
    const recordingInfo = await getRecordingInfo(callId, accessToken);
    
    if (recordingInfo.status !== 'Available' || !recordingInfo.url) {
      console.log(`⏭️ Recording not available for ${callId}`);
      return null;
    }
    
    console.log(`🔽 Downloading audio from: ${recordingInfo.url}`);
    
    // Download from the actual URL (no auth needed for this URL)
    const response = await axios.get(recordingInfo.url, {
      responseType: 'stream',
      timeout: 30000
    });
    
    // Use only first 3 characters of broker_id like the original script
    const shortBrokerId = brokerId.slice(0, 3);
    const filename = `${shortBrokerId}_${callId}.wav`;
    const filePath = path.join(audioDir, filename);
    
    // Create write stream and pipe the response
    const writer = fs.createWriteStream(filePath);
    response.data.pipe(writer);
    
    // Wait for download to complete
    await new Promise((resolve, reject) => {
      writer.on('finish', resolve);
      writer.on('error', reject);
    });
    
    console.log(`✅ Downloaded: ${filename}`);
    return { filePath, filename };
    
  } catch (error) {
    console.error(`❌ Failed to download ${callId}: ${error.message}`);
    return null;
  }
}

async function getAccessToken() {
  try {
    const params = new URLSearchParams();
    params.append('grant_type', 'client_credentials');
    params.append('client_id', process.env.NET2PHONE_CLIENT_ID);
    params.append('client_secret', process.env.NET2PHONE_CLIENT_SECRET);

    const response = await axios.post(
      config.api_config.net2phone.base_url + config.api_config.net2phone.token_endpoint,
      params,
      { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } }
    );

    return response.data.access_token;
  } catch (error) {
    throw new Error(`Failed to get access token: ${error.response?.data?.error || error.message}`);
  }
}

async function main() {
  console.log('🔽 Starting audio download (limited to 5 for testing)...');
  
  // Get access token first
  console.log('📡 Getting access token...');
  const accessToken = await getAccessToken();
  console.log('✅ Access token obtained');
  
  // Load call IDs from Excel files
  const callsWithRecordings = [];
  
  const xlsxFiles = fs.readdirSync(callIdsDir).filter(f => f.endsWith('.xlsx'));
  
  for (const xlsxFile of xlsxFiles) {
    if (xlsxFile === 'all_calls.xlsx') continue;
    
    const filePath = path.join(callIdsDir, xlsxFile);
    const workbook = xlsx.readFile(filePath);
    const sheetName = workbook.SheetNames[0];
    const data = xlsx.utils.sheet_to_json(workbook.Sheets[sheetName]);
    
    // Filter calls that have recording URLs
    data.forEach(row => {
      if (row.recording_url && row.recording_url.trim()) {
        callsWithRecordings.push({
          call_id: row.call_id,
          broker_id: row.broker_id || row.from_username || 'unknown',
          recording_url: row.recording_url,
          from_name: row.from_name || '',
          to_number: row.to_number || '',
          from_number: row.from_number || '',
          duration: row.duration || 0,
          start_time: row.start_time || ''
        });
      }
    });
  }
  
  console.log(`📊 Found ${callsWithRecordings.length} calls with recordings`);
  
  // Limit to test_limit for testing
  const testLimit = config.execution.test_limit || 5;
  const callsToProcess = callsWithRecordings.slice(0, testLimit);
  
  console.log(`🧪 Processing ${callsToProcess.length} calls for testing`);
  
  const downloadedFiles = [];
  
  for (let i = 0; i < callsToProcess.length; i++) {
    const call = callsToProcess[i];
    console.log(`\n📊 Progress: ${i + 1}/${callsToProcess.length}`);
    
    const result = await downloadAudio(
      call.recording_url,
      call.call_id,
      call.broker_id,
      accessToken
    );
    
    if (result) {
      downloadedFiles.push({
        ...call,
        filename: result.filename,
        file_path: result.filePath
      });
    }
    
    // Rate limiting
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  
  console.log(`\n✅ Audio download completed!`);
  console.log(`📊 Summary:`);
  console.log(`   - Downloaded: ${downloadedFiles.length}`);
  console.log(`   - Failed: ${callsToProcess.length - downloadedFiles.length}`);
  console.log(`   - Total: ${callsToProcess.length}`);
  
  return downloadedFiles.length > 0;
}

// Run if called directly
if (require.main === module) {
  main().then(success => {
    process.exit(success ? 0 : 1);
  });
}

module.exports = { main };

File: stages/2_download_recordings/download_recordings.py
#!/usr/bin/env python3

import os
import json
import pandas as pd
import requests
from pathlib import Path
import time

# Load configuration
config_path = os.environ.get('CONFIG_PATH', '../../config/pipeline_config.json')
with open(config_path, 'r') as f:
    config = json.load(f)

# Setup directories
base_dir = Path(config['directories']['base_dir'])
call_ids_dir = base_dir / 'output' / 'call_ids'
audio_dir = base_dir / 'output' / 'audio'
audio_dir.mkdir(exist_ok=True)

def download_audio(recording_url, call_id, broker_id):
    """Download audio file from recording URL"""
    try:
        print(f"🔽 Downloading audio for call {call_id}...")
        
        response = requests.get(recording_url, timeout=30)
        response.raise_for_status()
        
        filename = f"{broker_id}_{call_id}.wav"
        file_path = audio_dir / filename
        
        with open(file_path, 'wb') as f:
            f.write(response.content)
        
        print(f"✅ Downloaded: {filename}")
        return str(file_path), filename
        
    except Exception as e:
        print(f"❌ Failed to download {call_id}: {str(e)}")
        return None, None

def main():
    print("🔽 Starting audio download (limited to 5 for testing)...")
    
    # Load call IDs from Excel files
    calls_with_recordings = []
    
    for xlsx_file in call_ids_dir.glob('*.xlsx'):
        if xlsx_file.name == 'all_calls.xlsx':
            continue
            
        df = pd.read_excel(xlsx_file)
        
        # Filter calls that have recording URLs
        for _, row in df.iterrows():
            if pd.notna(row.get('recording_url')) and row['recording_url'].strip():
                calls_with_recordings.append({
                    'call_id': row['call_id'],
                    'broker_id': row.get('broker_id', row.get('from_username', 'unknown')),
                    'recording_url': row['recording_url'],
                    'from_name': row.get('from_name', ''),
                    'to_number': row.get('to_number', ''),
                    'from_number': row.get('from_number', ''),
                    'duration': row.get('duration', 0)
                })
    
    print(f"📊 Found {len(calls_with_recordings)} calls with recordings")
    
    # Limit to test_limit for testing
    test_limit = config['execution'].get('test_limit', 5)
    calls_to_process = calls_with_recordings[:test_limit]
    
    print(f"🧪 Processing {len(calls_to_process)} calls for testing")
    
    downloaded_files = []
    
    for i, call in enumerate(calls_to_process, 1):
        print(f"\n📊 Progress: {i}/{len(calls_to_process)}")
        
        file_path, filename = download_audio(
            call['recording_url'],
            call['call_id'],
            call['broker_id']
        )
        
        if file_path:
            downloaded_files.append({
                'call_id': call['call_id'],
                'broker_id': call['broker_id'],
                'filename': filename,
                'file_path': file_path,
                'from_name': call['from_name'],
                'to_number': call['to_number'],
                'from_number': call['from_number'],
                'duration': call['duration']
            })
        
        # Rate limiting
        time.sleep(1)
    
    print(f"\n✅ Audio download completed!")
    print(f"📊 Summary:")
    print(f"   - Downloaded: {len(downloaded_files)}")
    print(f"   - Failed: {len(calls_to_process) - len(downloaded_files)}")
    print(f"   - Total: {len(calls_to_process)}")

if __name__ == "__main__":
    main()